src/foam/dao/history/HistoryDAO.js

```diff
@@ -128,34 +129,46 @@ foam.CLASS({
     {
       name: 'put_',
       javaCode: `
-        Subject subject = (Subject) x.get("subject");
-        User user = subject.getUser();
-        User agent = subject.getRealUser();
-        FObject current = this.find_(x, obj);
-    
+      Subject subject = (Subject) x.get("subject");
+      User user = subject.getUser();
+      User agent = subject.getRealUser();
+      FObject current = this.find_(x, obj);
+      Object objectId = ((PropertyInfo) obj.getClassInfo().getAxiomByName("id")).f(obj);
+      boolean isCreate = objectId == null || 
+                          current == null ||
+                          SafetyUtil.isEmpty(String.valueOf(objectId)) ||
+                          new Long(0L).equals(objectId);
+      if ( isCreate ) {
+        // do "put" first if it is "create" action.
+        FObject persistObject = super.put_(x, obj);
+        objectId = ((PropertyInfo) persistObject.getClassInfo().getAxiomByName("id")).f(obj);
```

Comment:
Why are you not taking the objectId from the persistObject?  I though the issue is that the argument obj may not yet have an Id?

Reply:
oh, just a copy and paste mistake. I will update